src/Razor/test/Microsoft.AspNetCore.Razor.VSCode.FunctionalTest/tests/Hover2_2.test.ts

```diff
@@ -0,0 +1,158 @@
+        }
+
+        hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(0, 8));
+
+        assert.ok(hoverResult, 'Should have a hover result for asp-for');
+        if (hoverResult) {
+            assert.equal(hoverResult.length, 1, 'Something else may be providing hover results');
+
+            const envResult = hoverResult[0];
+            const expectedRange = new vscode.Range(
+                new vscode.Position(0, 8),
+                new vscode.Position(0, 8));
+            assert.deepEqual(envResult.range, expectedRange, 'asp-for should be selected');
+            const mStr = envResult.contents[0] as vscode.MarkdownString;
+            assert.ok(mStr.value.includes('InputTagHelper.**For**'), `InputTagHelper.For not included in '${mStr.value}'`);
+        }
+
+        hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(0, 19));
+
+        assert.ok(hoverResult, 'Should have a hover result for class');
+        if (hoverResult) {
+            assert.equal(hoverResult.length, 1, 'Something else may be providing hover results');
+
+            const result = hoverResult[0];
+            const expectedRange = new vscode.Range(
+                new vscode.Position(0, 19),
+                new vscode.Position(0, 24));
+            assert.deepEqual(result.range, expectedRange, 'class should be selected');
+            const mStr = result.contents[0] as vscode.MarkdownString;
+            assert.ok(mStr.value.includes('class'), `class not included in ${mStr.value}`);
+        }
+    });
+
+    // MvcWithComponents doesn't find TagHelpers because of test setup foibles.
+    test('Can perform hovers on TagHelpers', async () => {
+        const firstLine = new vscode.Position(0, 0);
+        await editor.edit(edit => edit.insert(firstLine, '<environment class="someName"></environment>\n'));
+        const hoverResult = await vscode.commands.executeCommand<vscode.Hover[]>(
+            'vscode.executeHoverProvider',
+            cshtmlDoc.uri,
+            new vscode.Position(0, 3));
+        const expectedRange = new vscode.Range(
+            new vscode.Position(0, 3),
+            new vscode.Position(0, 3));
+
+        assert.ok(hoverResult, 'Should have a hover result for EnvironmentTagHelper');
+        if (hoverResult) {
+            assert.equal(hoverResult.length, 1, 'Something else may be providing hover results');
+
+            const envResult = hoverResult[0];
+            assert.deepEqual(envResult.range, expectedRange, 'TagHelper range should be <environment>');
+            const mStr = envResult.contents[0] as vscode.MarkdownString;
+            assert.ok(mStr.value.includes('**EnvironmentTagHelper**'), `EnvironmentTagHelper not included in '${mStr.value}'`);
+        }
+    });
+});
```

Comment:
Maybe I overlooked it but do we have a test that verifies hovering over TagHelper attributes?

Reply:
It's buried in `Can perform hovers on TagHelpers that need attributes`. I'm modifying the name to reflect this.